src/fastpath/fastpath.c

```diff
@@ -6,15 +6,254 @@
+#ifdef ENABLE_SMP_SUPPORT
+#ifdef CONFIG_HAVE_FPU
+            if (nativeThreadUsingFPU(dest)) {
+                slowpath(SysSend);
+                UNREACHABLE();
+            }
+#endif
+#endif
+        } else {
+            sc = dest->tcbSchedContext;
+        }
+
+        /* Only fastpath signal to threads which will not become the new highest prio thread on the
+            core of their SC, even if the currently running thread on the core is the idle thread. */
+        if (NODE_STATE_ON_CORE(ksCurThread, sc->scCore)->tcbPriority < dest->tcbPriority) {
+            slowpath(SysSend);
+        }
+
+        /* Simplified schedContext_resume that does not change state and reverts to the
+         * slowpath in cases where the SC does not have sufficient budget, as this case
+         * adds extra scheduler logic. Normally, this is done after donation of SC
+         * but after tweaking it, I don't see anything executed in schedContext_donate
+         * that will affect the conditions of this check */
+        if (sc->scRefillMax > 0) {
+            if (!(refill_ready(sc) && refill_sufficient(sc, 0))) {
+                slowpath(SysSend);
+                UNREACHABLE();
+            }
+            schedulable = true;
+        }
+
+        /* Check if signal is cross-core or cross-domain */
+        if (ksCurDomain != dest->tcbDomain SMP_COND_STATEMENT( || sc->scCore != getCurrentCPUIndex())) {
+            crossnode = true;
+        }
+
+        /*  Point of no return */
+#ifdef CONFIG_BENCHMARK_TRACK_KERNEL_ENTRIES
+        ksKernelEntry.is_fastpath = true;
+#endif
+        /* Dequeue dest from the notification queue */
+        ntfn_queue_dequeue_fp(dest, ntfnPtr);
+
+        /* Wake up the signalled thread and tranfer badge */
+        setRegister(dest, badgeRegister, badge);
+        thread_state_ptr_set_tsType_np(&dest->tcbState, ThreadState_Running);
+
+        /* Donate SC if necessary. The checks for this were already done before
+         * the point of no return */
+        maybeDonateSchedContext_fp(dest, sc);
+
+        /* Left this in the same form as the slowpath. Not sure if optimal */
+        if (sc_sporadic(dest->tcbSchedContext)) {
+            assert(dest->tcbSchedContext != NODE_STATE(ksCurSC));
+            if (dest->tcbSchedContext != NODE_STATE(ksCurSC)) {
+                refill_unblock_check(dest->tcbSchedContext);
+            }
+        }
+
+        /* If dest was already not schedulable prior to the budget check
+         * the slowpath doesn't seem to do anything special besides just not
+         * not scheduling the dest thread. */
+        if (schedulable) {
+            if (NODE_STATE(ksCurThread)->tcbPriority < dest->tcbPriority || crossnode) {
+                SCHED_ENQUEUE(dest);
+            } else {
+                SCHED_APPEND(dest);
+            }
+        }
+
+        restore_user_context();
+        UNREACHABLE();
```

Comment:
This part is exactly the same as for `NtfnState_Idle` idle case with bound nf, except for the `cancelIPC_fp` call in the idle case. Please share the code as much as possible instead of duplicating it. Perhaps move it to separate function and call it in both cases? Or move this block out of the switch if you don't want a separate function.

Reply:
I didn't put the the stuff before the point of no return in functions because the repeated code involves slowpath transitions and I think Gerwin mentioned this should stay out of functions. I could put the duplicated stuff after the point of no return into a function though, I just didn't because I thought it wasn't a lot.

Reply:
It's 1/3rd of the whole signal fastpath code, I would think that's worth it. See [this commit](https://github.com/indanz/seL4/commit/4270ae675526bc0290edbc17ea682a8c3b24fbf3) for the difference. I tested the code on our platform, it seems to work with `KernelSignalFastpath` enabled. (I forked your branch to [here](https://github.com/seL4/seL4/compare/master...Indanz:seL4:signal_fp3))

Reply:
Point taken, briefly looking at your fork, it looks close to being equivalent and will reduce duplicated code for a couple of extra checks.